LSP Find References Feature

niteskum opened the pull request.

Commit:
LSP Find References Feature

niteskum commented:
@shubhsnov  Please review

Commit:
Addressed bugs

niteskum commented:
Find References UI:
![image](https://user-images.githubusercontent.com/38075523/55938602-5508de00-5c59-11e9-8cb0-4a1b5c224048.png)

shubhsnov commented:
@niteskum I've still got to check the functional aspects, but I've given a few code nits that I came across on first glance.

niteskum commented:
Find References UI:

On Right Click:
![image](https://user-images.githubusercontent.com/38075523/55938676-81245f00-5c59-11e9-96a4-bdbd4da474da.png)

References Results UI:
![image](https://user-images.githubusercontent.com/38075523/55938723-9c8f6a00-5c59-11e9-9b15-4c8e40888e6b.png)

Commit:
Addressed review comments

Commit:
Addressed review comments

Commit:
Changes to send data only once per 24 hours

Commit:
Adding Unit Test cases and supporting files

Commit:
Merge branch 'snchit_fix_for_send_event_every_24_hours' of https://github.com/niteskum/brackets into lspFindReferences

Commit:
changed string

shubhsnov commented:
@niteskum The string should get highlighted as in search results. Is there any particular reason for why this is not happening?
![image](https://user-images.githubusercontent.com/16770513/56058211-5da70480-5d7e-11e9-9dd1-2bb0b3490998.png)

Commit:
fixed review comments

Commit:
Addressed review comments

niteskum commented:
> @niteskum The string should get highlighted as in search results. Is there any particular reason for why this is not happening?
> ![image](https://user-images.githubusercontent.com/16770513/56058211-5da70480-5d7e-11e9-9dd1-2bb0b3490998.png)

fixed.

shubhsnov commented on src/features/keyboard.json:
```diff
@@ -0,0 +1,13 @@
+{
```
Creating a new keyboard.json is something we do for extensions. Since FindReferences is a core feature, you should add this to the main [keyboard.json](https://github.com/adobe/brackets/blob/master/src/base-config/keyboard.json).

Change the key to "cmd.findAllReferences".

Additionally, add the menu entries under [DefaultMenus.js](https://github.com/adobe/brackets/blob/master/src/command/DefaultMenus.js) instead of in FindReferencesManager.js

Also, you'll need to add it at two places, one is the context menu [here](https://github.com/adobe/brackets/blob/master/src/command/DefaultMenus.js#L281) and secondly in the AppBarMenu [here](https://github.com/adobe/brackets/blob/master/src/command/DefaultMenus.js#L138), below FIndInFiles

shubhsnov commented on src/search/SearchResultsView.js:
```diff
@@ -72,14 +72,16 @@ define(function (require, exports, module) {
      * @param {SearchModel} model The model that this view is showing.
      * @param {string} panelID The CSS ID to use for the panel.
      * @param {string} panelName The name to use for the panel, as passed to WorkspaceManager.createBottomPanel().
+     * @param {string} type type to indentify if it is reference search or string match serach
```
You still need to fix this.

shubhsnov commented on src/features/FindReferencesManager.js:
```diff
@@ -0,0 +1,148 @@
+        ),
+        removeFindReferencesProvider = _providerRegistrationHandler.removeProvider.bind(_providerRegistrationHandler);
+
+    var SHOW_FIND_REFERENCES_CMD_ID   = "showReferences",
```
Since this is a core feature, move this to [Commands.js](https://github.com/adobe/brackets/blob/master/src/command/Commands.js#L90)

Change string to `"cmd.findAllReferences"`. This will allow you to use it in DefaultMenus.js.

shubhsnov commented on src/features/keyboard.json:
```diff
@@ -0,0 +1,13 @@
+        {
+            "key": "Ctrl-Shift-R", "platform": "win"
+        },
+        {
```
If the keys are same for all platforms, then just one key should work for all platforms. Refer [this](https://github.com/adobe/brackets/blob/master/src/base-config/keyboard.json).

```
{
     "key": "Ctrl-Shift-R"
}
```

shubhsnov commented on src/languageTools/DefaultProviders.js:
```diff
@@ -404,8 +405,82 @@ define(function (require, exports, module) {
+                            var filePath = PathConverters.uriToPath(element.uri);
+                            DocumentManager.getDocumentForPath(filePath)
+                                .done(function(doc) {
+                                    var startRange = {line: element.range.start.line, ch: element.range.start.character};
```
Move this transformation logic to another function so this can be used outside of the DefaultProviders, if necessary.

shubhsnov commented:
Apart from a few refactoring changes, this look good to me.

Commit:
Addressed review comments

niteskum commented:
> Apart from a few refactoring changes, this looks good to me.

@shubhsnov  Addressed all review comments.
for "Find All References" I am using "Ctrl-Shift-K"  shortcut as "Ctrl-Shift-R" is already used for reloadLivePreview.

shubhsnov commented on src/languageTools/DefaultProviders.js:
```diff
@@ -404,8 +405,96 @@ define(function (require, exports, module) {
+            }).done(function(msgObj){
+                    if(msgObj && msgObj.length) {
+                        msgObj.cursorPos = pos;
+                        serverRespToSerachModelFormat(msgObj)
```
Nit: Typo "`serverRespToSearchModelFormat`"

shubhsnov commented on src/features/FindReferencesManager.js:
```diff
@@ -0,0 +1,140 @@
+    function _getReferences(provider, hostEditor, pos) {
+        var result = new $.Deferred();
+
+        if(!provider) {
```
Since you are using _resultsView and AppInit takes some time to dispatch AppReady (this should be **htmlReady** right? as in [FindinFilesUI.js](https://github.com/adobe/brackets/blob/master/src/search/FindInFilesUI.js#L454)), you should put a null check for _resultsView here as well.

shubhsnov commented on src/features/FindReferencesManager.js:
```diff
@@ -0,0 +1,140 @@
+        searchModel.clear();
+    }
+
+    AppInit.appReady(function () {
```
This should be **htmlReady**, I think. For non-LSP based providers, DOM initialization is more than enough. See my comment above.

shubhsnov commented on src/extensions/default/PhpTooling/main.js:
```diff
@@ -61,7 +62,8 @@ define(function (require, exports, module) {
         lProvider,
-        jdProvider;
+        jdProvider,
+        refProvider;
```
Resolve the merge conflicts.

Commit:
Merge branch 'master' of https://github.com/adobe/brackets into lspFindReferences

narayani28 commented:
ctrl+k is used for quick docs and ctrl+shift+k is not very intuitive. is there any non used ctrl+<alphabet> available for usage?

shubhsnov commented:
LGTM 👍

Commit:
Changed Shortcut to Ctrl-M

Commit:
Changed the Shortcut to Shift-F12
